src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Package/Update/PackageUpdateIO.cs

```diff
@@ -181,12 +207,234 @@ public void UpdatePackageReference(PackageSpec updatedPackageSpec, IPackageUpdat
+    public async Task<IReadOnlyList<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>>> GetKnownVulnerabilitiesAsync(ILogger logger, CancellationToken cancellationToken)
+    {
+        IReadOnlyList<PackageSource>? auditSources = _sourceProvider.LoadAuditSources()?.Where(s => s.IsEnabled).ToList();
+        if (auditSources is null || auditSources.Count == 0)
+        {
+            auditSources = _enabledSources;
+        }
+
+        var tasks = new List<Task<GetVulnerabilityInfoResult?>>(auditSources.Count);
+        foreach (var auditSource in auditSources)
+        {
+            tasks.Add(Task.Run(async () =>
+            {
+                var sourceRepository = Repository.Factory.GetCoreV3(auditSource.Source);
+                var vulnerabilityResource = await sourceRepository.GetResourceAsync<IVulnerabilityInfoResource>(cancellationToken);
+                if (vulnerabilityResource is not null)
+                {
+                    var vulnerabilities = await vulnerabilityResource.GetVulnerabilityInfoAsync(_sourceCacheContext, logger, cancellationToken);
+                    return vulnerabilities;
+                }
+                return null;
+            }, cancellationToken));
+        }
+
+        List<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>> allVulnerabilities = new();
+        foreach (var task in tasks)
+        {
+            var result = await task;
+            if (result is not null)
+            {
+                if (result.KnownVulnerabilities?.Count > 0)
+                {
+                    foreach (var vulnDict in result.KnownVulnerabilities)
+                    {
+                        allVulnerabilities.Add(vulnDict);
+                    }
+                }
+            }
+        }
+
+        return allVulnerabilities;
+    }
+
+    /// <inheritdoc cref="IPackageUpdateIO.GetNonVulnerableAsync(string, NuGetVersion, ILogger, IReadOnlyList{IReadOnlyDictionary{string, IReadOnlyList{PackageVulnerabilityInfo}}}, CancellationToken)"/>
+    public async Task<NuGetVersion?> GetNonVulnerableAsync(
+        string packageId,
+        NuGetVersion minVersion,
+        ILogger logger,
+        IReadOnlyList<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>> knownVulnerabilities,
+        CancellationToken cancellationToken)
+    {
+        var sources = GetSourcesForPackage(packageId);
+        var lookups = new Task<NuGetVersion?>[sources.Count];
+        for (int source = 0; source < sources.Count; source++)
+        {
+            SourceRepository sourceRepository = sources[source];
+            // If package source is a local folder feed, it might not actually be async
+            lookups[source] = Task.Run(() => FindLowestNonVulnerablePackageVersionAsync(sourceRepository, packageId, minVersion, knownVulnerabilities, logger, cancellationToken));
+        }
+
+        await Task.WhenAll(lookups);
+
+        NuGetVersion? lowestVersion = null;
```

Comment:
Could we rename to lowestNonVulnerableVersion? When I first read this, I thought you were literally looking for the lowest version.